src/main/java/tweeting/services/TwitterService.java

```diff
@@ -0,0 +1,107 @@
+
+    private static TwitterService instance;
+
+    private Twitter api;
+
+    private static final Logger logger = LoggerFactory.getLogger(TwitterService.class);
+
+    public static final int MAX_TWEET_LENGTH = CharacterUtil.MAX_TWEET_LENGTH; // To not expose Twitter4J
+
+    private TwitterService() {
+    }
+
+    public static TwitterService getInstance(TwitterOAuthCredentials auth) {
+        try {
+            if (instance == null) {
+                instance = new TwitterService();
+            }
+            ConsumerAPIKeys consumerAPIKeys = auth.getConsumerAPIKeys();
+            AccessTokenDetails accessTokenDetails = auth.getAccessTokenDetails();
+            ConfigurationBuilder configurationBuilder = new ConfigurationBuilder();
+            configurationBuilder.setDebugEnabled(true);
+            configurationBuilder.setJSONStoreEnabled(true); // Need in order to use getRawJSON
+            configurationBuilder.setOAuthConsumerKey(consumerAPIKeys.getConsumerAPIKey());
+            configurationBuilder.setOAuthConsumerSecret(consumerAPIKeys.getConsumerAPISecretKey());
+            configurationBuilder.setOAuthAccessToken(accessTokenDetails.getAccessToken());
+            configurationBuilder.setOAuthAccessTokenSecret(accessTokenDetails.getAccessTokenSecret());
+            TwitterFactory twitterFactory = new TwitterFactory(configurationBuilder.build());
+            instance.api = twitterFactory.getInstance();
+            return instance;
+        } catch (Exception e) {
+            logger.error(e.getMessage(), e); // Log error message in cause initialization fails
+            throw e;
+        }
+    }
+
+    public static TwitterService getInstance() {
+        if (instance == null) {
+            instance = new TwitterService();
+            TwitterFactory.getSingleton();
+            logger.warn("TwitterService has been instantiated with no Twitter credentials. Please call getInstance" +
+                    "with TwitterOAuthCredentials parameter to set credentials to support calls to Twitter.");
+        }
+        return instance;
+    }
+
+    public List<Status> getTweets() throws TwitterServiceException {
+        try {
+            return api.getHomeTimeline();
+        } catch (TwitterException te) {
+            throw new TwitterServiceException(te);
+        } catch (Exception e) {
+            logger.error(e.getMessage(), e);
+            return null;
+        }
+    }
+
+    public Status postTweet(String message) throws TwitterServiceException {
+        // Prelim checks (avoid calling to Twitter if unnecessary)
+        if (message == null) {
+            TwitterServiceException e = new TwitterServiceException("Request is missing message parameter.",
+                    new NullPointerException("message"));
+            throw e;
+        }
+        if (StringUtils.isBlank(message)) {
+            TwitterServiceException e = new TwitterServiceException("Message parameter is blank.",
```

Comment:
instead of error codes, it would be more straightforward to use different exceptions. One for unknown exceptions and one for other exceptions.